### DISCUSSION TITLE
How to use `makeWrapper` in a startup script?

### DISCUSSION THREAD

**Discussion by poperigby** [numtide/devshell#333] [Q&A]
When I add `makeWrapper` to my list of packages, the startup script isn't able to find the `makeWrapper` bash function.

```nix
devshells.default = {
  packagesFrom = [ config.packages.default ];

  packages = with pkgs; [
    gdb

    libsForQt5.wrapQtAppsHook
    makeWrapper
    bashInteractive
  ];

  devshell.startup.wrapQt.text = # bash
    ''
      # Set the environment variables that unpatched Qt apps expect
      bashdir=$(mktemp -d)
      makeWrapper "$(type -p bash)" "$bashdir/bash" "''${qtWrapperArgs[@]}"
      exec "$bashdir/bash"
    '';
};
```

**Comment by onemoresuza** [comment]
> When I add `makeWrapper` to my list of packages, the startup script isn't able to find the `makeWrapper` bash function.
>
> ```nix
> devshells.default = {
>   packagesFrom = [ config.packages.default ];
>
>   packages = with pkgs; [
>     gdb
>
>     libsForQt5.wrapQtAppsHook
>     makeWrapper
>     bashInteractive
>   ];
>
>   devshell.startup.wrapQt.text = # bash
>     ''
>       # Set the environment variables that unpatched Qt apps expect
>       bashdir=$(mktemp -d)
>       makeWrapper "$(type -p bash)" "$bashdir/bash" "''${qtWrapperArgs[@]}"
>       exec "$bashdir/bash"
>     '';
> };
> ```

The `pkgs.makeWrapper` derivation does not provide a `makeWrapper` file
under `$out/bin`, since it provides only a setup-hook, thus, there's no
makeWrapper to be used on the startup script. 

To make it available to your startup script, you'd have to source it
like so:

```bash
source "${pkgs.makeWrapper}/nix-support/setup-hook"
```

But do beware of sourcing setup-hooks: They're made to work only
under the nixpkgs build environment, having no guarantees that they work
outside of it.

**Reply by poperigby** [reply]
Thank you. That seems to just infinitely loop for some reason though :cry:.
